astra/src/main/java/com/slack/astra/zipkinApi/ZipkinService.java

```diff
@@ -242,6 +281,93 @@ public HttpResponse getTraceByTraceId(
+    }
     return HttpResponse.of(HttpStatus.OK, MediaType.JSON_UTF_8, output);
   }
+
+  @VisibleForTesting
+  protected static byte[] compressJsonData(String jsonData) throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    try (GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream)) {
+      gzipOutputStream.write(jsonData.getBytes(StandardCharsets.UTF_8));
+    }
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  @VisibleForTesting
+  protected static String decompressJsonData(byte[] compressedData) {
+    assert compressedData != null && compressedData.length > 0;
+
+    try (GZIPInputStream gzipInputStream =
+        new GZIPInputStream(new ByteArrayInputStream(compressedData))) {
+      return new String(gzipInputStream.readAllBytes(), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      LOG.error("Error decompressing JSON data", e);
+      throw new RuntimeException("Failed to decompress JSON data", e);
+    }
+  }
+
+  protected String retrieveDataFromS3(String traceId, long dataFreshnessInMinutes) {
+    assert traceId != null && !traceId.isEmpty();
+
+    try {
+      // If data is fresh skip S3 retrieval
+      long currentTime = Instant.now().toEpochMilli();
+
+      HeadObjectResponse response =
+          blobStore.getFileMetadata(
+              String.format("%s/%s/traceData.json.gz", TRACE_CACHE_PREFIX, traceId));
+      long lastModified = response.lastModified().toEpochMilli();
+      if (currentTime - lastModified < dataFreshnessInMinutes * 60 * 1000) {
+        return null; // Data is still getting updated, check on cache and live nodes
+      }
+
+      // Retrieve the compressed trace data from S3
+      java.nio.file.Path tempDir = Files.createTempDirectory("");
+      blobStore.download(String.format("%s/%s", TRACE_CACHE_PREFIX, traceId), tempDir);
+
+      // Decompress the JSON data
+      byte[] compressedData = Files.readAllBytes(tempDir.resolve("traceData.json.gz"));
+      String jsonData = decompressJsonData(compressedData);
+
+      LOG.info("Retrieved and decompressed trace data from S3 for traceId={}", traceId);
+      // Process the jsonData as needed
+      return jsonData;
+
+    } catch (Exception e) {
+      LOG.error("Error retrieving trace data from S3 for traceId={}", traceId, e);
+    }
+    return null;
+  }
+
+  protected void saveDataToS3(String traceId, String output) {
+    assert traceId != null && !traceId.isEmpty();
+    assert output != null && !output.isEmpty();
+
+    try {
+      // Compress the JSON data using GZIP
+      byte[] compressedData = compressJsonData(output);
+
+      // Create a temporary directory to store the trace data
+      java.nio.file.Path tempDir = Files.createTempDirectory(traceId);
+      java.nio.file.Path traceFile = tempDir.resolve("traceData.json.gz");
+      Files.write(traceFile, compressedData);
+
+      // Upload the compressed trace data to S3
```

Comment:
I think we need to handle the concurrent write problem where two writers are trying to upload the data at the same time. Consider writing to a traceId-temp prefix in S3 and then renaming it so we handle the concurrent writer issue.